tripy/tripy/flat_ir/ops/constant.py

```diff
@@ -44,34 +43,46 @@ def to_mlir(self, operands):
+                )
+
+            # Workaround (#208): bools are represented as i1 in MLIR-TRT but they cannot be used for DenseElementsAttr
+            # so we have to represent them as ints and then cast the result
+            if self.outputs[0].dtype == datatype.bool:
+                # need to use memoryview.cast to ensure that the view will be flattened
+                int_memref = self.data.runtime_client.create_memref(
+                    array.array("i", memoryview(memref_value).cast("b").tolist()),
+                    shape=self.data.shape,
+                    dtype=mlir_utils.convert_tripy_dtype_to_runtime_dtype(datatype.int32),
+                    device=None,
+                )
+                attr = ir.DenseElementsAttr.get(
+                    array=int_memref, type=mlir_utils.get_mlir_dtype(datatype.int32), shape=self.data.shape
+                )
+                cast_output = mlir_utils.make_mlir_tensor(datatype.bool, self.data.shape)
+                constant_op = stablehlo.ConstantOp(attr)
+                return [stablehlo.ConvertOp(result=cast_output, operand=constant_op)]
 
-        # Workaround (#208): bools are represented as i1 in MLIR-TRT but they cannot be used for DenseElementsAttr
-        # so we have to represent them as ints and then cast the result
-        if self.outputs[0].dtype == datatype.bool:
-            # need to use memoryview.cast to ensure that the view will be flattened
-            int_memref = self.data.runtime_client.create_memref(
-                array.array("i", memoryview(memref_value).cast("b").tolist()),
-                shape=self.data.shape,
-                dtype=mlir_utils.convert_tripy_dtype_to_runtime_dtype(datatype.int32),
-                device=None,
-            )
             attr = ir.DenseElementsAttr.get(
-                array=int_memref, type=mlir_utils.get_mlir_dtype(datatype.int32), shape=self.data.shape
+                array=memref_value, type=mlir_utils.get_mlir_dtype(self.outputs[0].dtype), shape=self.data.shape
             )
-            cast_output = mlir_utils.make_mlir_tensor(datatype.bool, self.data.shape)
-            constant_op = stablehlo.ConstantOp(attr)
-            return [stablehlo.ConvertOp(result=cast_output, operand=constant_op)]
+        else:
+            out_dtype = self.outputs[0].dtype
 
-        assert self.data.dtype == self.outputs[0].dtype
-        attr = ir.DenseElementsAttr.get(
-            array=memref_value, type=mlir_utils.get_mlir_dtype(self.outputs[0].dtype), shape=self.data.shape
-        )
+            def to_attrs(data):
+                if isinstance(data, numbers.Number):
+                    return [mlir_utils.get_mlir_scalar_attr(out_dtype, data)]
+                attrs = []
+                for element in data:
+                    attrs.extend(to_attrs(element))
+                return attrs
+
+            attr = ir.DenseElementsAttr.get(attrs=to_attrs(self.data.data()), type=self.outputs[0].to_mlir())
```

Comment:
@jhalakpatel This is the logic that converts a python list to MLIR constant:

- There is no cast because we are creating the MLIR attr of target dtype directly.
- `convert_list_to_array` is not called here, MLIR attr accpets python numbers directly, there should be no perf impact.